packages/response-ops/rule_form/src/hooks/use_rule_form_steps.tsx

```diff
@@ -0,0 +1,300 @@
+            status: actionsStatus,
+            children: (
+              <>
+                <RuleActions />
+                <EuiSpacer />
+                <EuiHorizontalRule margin="none" />
+              </>
+            ),
+          }
+        : null,
+      [RuleFormStepId.DETAILS]: {
+        title: RULE_FORM_PAGE_RULE_DETAILS_TITLE,
+        status: ruleDetailsStatus,
+        children: (
+          <>
+            <RuleDetails />
+            <EuiSpacer />
+            <EuiHorizontalRule margin="none" />
+          </>
+        ),
+      },
+    }),
+    [ruleDefinitionStatus, canReadConnectors, actionsStatus, ruleDetailsStatus]
+  );
+
+  const stepOrder: RuleFormStepId[] = useMemo(
+    () => STEP_ORDER.filter((stepId) => steps[stepId]),
+    [steps]
+  );
+
+  return { steps, stepOrder };
+};
+
+interface RuleFormVerticalSteps {
+  steps: EuiStepsProps['steps'];
+}
+
+export const useRuleFormSteps: () => RuleFormVerticalSteps = () => {
+  // Track steps that the user has interacted with and then focused away from
+  const [touchedSteps, setTouchedSteps] = useState<Record<RuleFormStepId, boolean>>(
+    STEP_ORDER.reduce(
+      (result, stepId) => ({ ...result, [stepId]: false }),
+      {} as Record<RuleFormStepId, boolean>
+    )
+  );
+
+  const reportOnBlur = useCallback(
+    (stepId: RuleFormStepId, element: React.JSX.Element) => (
+      <div
+        data-test-subj={`ruleFormStep-${stepId}-reportOnBlur`}
+        onBlur={() =>
+          !touchedSteps[stepId] &&
+          setTouchedSteps({
+            ...touchedSteps,
+            [stepId]: true,
+          })
+        }
+      >
+        {element}
+      </div>
+    ),
+    [touchedSteps]
+  );
+
+  const { steps, stepOrder } = useCommonRuleFormSteps({ touchedSteps });
+
+  const mappedSteps = useMemo(() => {
+    return stepOrder
+      .map((stepId) => {
+        const step = steps[stepId];
+        return step
+          ? {
+              ...step,
+              children: reportOnBlur(stepId, step.children),
+            }
+          : null;
+      })
+      .filter(Boolean) as EuiStepsProps['steps'];
+  }, [steps, stepOrder, reportOnBlur]);
+
+  return { steps: mappedSteps };
+};
+
+interface RuleFormHorizontalSteps {
+  steps: EuiStepsHorizontalProps['steps'];
+  currentStepComponent: React.ReactNode;
+  goToNextStep: () => void;
+  goToPreviousStep: () => void;
+  hasNextStep: boolean;
+  hasPreviousStep: boolean;
+}
+export const useRuleFormHorizontalSteps: () => RuleFormHorizontalSteps = () => {
+  const [currentStep, setCurrentStep] = useState<RuleFormStepId>(STEP_ORDER[0]);
+  const [touchedSteps, setTouchedSteps] = useState<Record<RuleFormStepId, boolean>>(
+    STEP_ORDER.reduce(
+      (result, stepId) => ({ ...result, [stepId]: false }),
+      {} as Record<RuleFormStepId, boolean>
+    )
+  );
+
+  const { steps, stepOrder } = useCommonRuleFormSteps({
+    touchedSteps,
+    currentStep,
+  });
+
+  const hasNextStep = useMemo(() => {
```

Comment:
nit: Wdyt of exporting the logic on a function (in the same file) to avoid making the hook too big and make it more readable? For example:

```
const hasNextStep = useMemo(() => _hasNextStep(currentStep, stepOrder), [//deps])
```

Reply:
I think moving the logic out of the hook would actually make it more difficult to read, but I was able to make this code block more compact like this:

```js
const currentStepIndex = useMemo(() => stepOrder.indexOf(currentStep), [currentStep, stepOrder]);
  const hasNextStep = useMemo(
    () => currentStep && currentStepIndex < stepOrder.length - 1,
    [currentStepIndex, currentStep, stepOrder]
  );
  const hasPreviousStep = useMemo(
    () => currentStep && currentStepIndex > 0,
    [currentStepIndex, currentStep]
  );
```